### PR TITLE
Remove failed vmss

### PR DESCRIPTION
**Pull request by nilsanderselde**
### Which issue this PR addresses:

Fixes [10485408](https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/10485408)

### What this PR does / why we need it:

- Cleans up VMSS after failed RP/Gateway deploy to prevent conflict when VMSS with same name is created on retry
- Refactors common logic out of RP/Gateway deployment
- Adds retries to Gateway deploy
- Gives one file a better name given its content and fixes capitalization of a few vars

### Test plan for issue:

Either find a way to force the deployment to fail in INT and ensure the VMSS is deleted, or just ensure that it at least does not _prevent_ successful deployment and we can see if it has an effect.

### Is there any documentation that needs to be updated for this PR?

If there are docs saying to manually delete the VMSS, yes.

### CODE REVIEW

**Review by m1kola — CHANGES_REQUESTED** [review]
I probably am missing some context (work item is light on details), but I think with this code we _might_ end up without VMs to run RP at all (region is going to be down). Doesn't sound like it is something we want to do.

**Comment by nilsanderselde** [comment]
>missing some context

Occasionally during RP deployment, the new VMSS goes into Failed state and the deployment exits. The old VMSS is still serving at this time. In order to retry the deployment, because the new VMSS will have the same name as the failed one, the failed one must currently be manually deleted (or else we get the Confict error). The intent here is to do so automatically—to clean up after a failed deployment. 

It should only delete VMSSs in Failed provisioning state; the old, serving VMSS should not be in Failed state.

You made me realize that we should check the name, though, to be absolutely sure we are not deleting any other VMSS which just happens to be in Failed state (even though that shouldn't be the case).

**Comment by nilsanderselde** [comment]
@m1kola I switched over to use Get instead of List, since we should only be deleting one specific VMSS.

File: pkg/deploy/deploy_rp.go
Comment (jewzaam) [inline pkg/deploy/deploy_rp.go:99]:
```suggestion
				return innerErr
			}
			continue
```

And bump the loop (which I cannot seem to put a suggestion on) to be `i < 3` to give this a chance to retry.  There are now 2 error conditions we need to account for so need at least 3 attempts at creating the VMSS.

File: pkg/deploy/deploy_rp.go
Comment (jewzaam) [inline pkg/deploy/deploy_rp.go:200]: Does this return `nil` for `err` if the VMSS doesn't exist? We need it to not fail if the VMSS does not exist or have this function return the vmss and the error so caller can determine what to do.

File: pkg/deploy/deploy_rp.go
Comment (jewzaam) [inline pkg/deploy/deploy_rp.go:205]: Use constant `ProvisioningStateFailed` from azure sdk instead of hard coded string.

**Review by jewzaam — CHANGES_REQUESTED** [review]

File: pkg/deploy/deploy_rp.go
Comment (nilsanderselde) [reply pkg/deploy/deploy_rp.go:99]: Point taken - I'm not sure that the above suggestion will work exactly, though. We still need to return `err` somewhere. I'll see what I can do.

File: pkg/deploy/deploy_rp.go
Comment (nilsanderselde) [reply pkg/deploy/deploy_rp.go:200]: It would return an error. The error text from the CLI is below:
>(ResourceNotFound) The Resource 'Microsoft.Compute/virtualMachineScaleSets/x' under resource group 'y' was not found. For more details please go to https://aka.ms/ARMResourceNotFoundFix

`rpRemoveFailedScaleset` would return the `ResourceNotFound` error. Then, `DeployRP` would log the `ResourceNotFound` error, and either continue (if there are still attempts left) or return the original deployment failure error.

File: pkg/deploy/deploy_rp.go
Comment (nilsanderselde) [reply pkg/deploy/deploy_rp.go]: Fixed

File: pkg/deploy/deploy_rp.go
Comment (nilsanderselde) [reply pkg/deploy/deploy_rp.go:105]: @jewzaam I've updated this in a way I think will work, if you want to take another look

File: pkg/deploy/deploy_rp.go
Comment (jewzaam) [inline pkg/deploy/deploy_rp.go:102]: This dependency on specific index, 1 and now 2, just.. it's wrong.  But it's not wrong enough for me to ding the PR I think.  Rather see an open loop and a set of conditions checked and exit if the conditions are met else continue with some cap on iterations.  Just had to write this down as e2e is still going and in case someone felt this needed addressed.  We have time.  But I'll approve.

**Review by jewzaam — DISMISSED** [review]

File: pkg/deploy/deploy_rp.go
Comment (nilsanderselde) [reply pkg/deploy/deploy_rp.go:102]: Yeah, using the specific `i` values is pretty nasty. There's got to be a better way.

I also have another concern: I'm not sure that retrying immediately after attempting to delete the failed VMSS will necessarily prevent the resource conflict. If it doesn't "go away" fast enough, there might be a race condition between deleting the resource and trying to recreate it. This is why I originally didn't retry, I just deleted the VMSS and returned the original error. I figured we could retry at the pipeline level, and just be spared the pain of manually deleting the VMSS.

Comment (jewzaam) [reply pkg/deploy/deploy_rp.go:102]: Maybe a condition if there's an existing VMSS and status is "Deleting" wait up to.. 10 minutes?

File: pkg/util/azureclient/mgmt/compute/virtualmachinesscalesets_addons.go
Comment (m1kola) [inline pkg/util/azureclient/mgmt/compute/virtualmachinesscalesets_addons.go:51]: No need to wrap this: upstram implementation which is embeded) matches this interface. See the comment about `VirtualMachineScaleSetsClient` vs `VirtualMachineScaleSetsClientAddons` above.

```suggestion
```

**Review by m1kola — COMMENTED** [review]
Looks good in general, but we need to clean up the client changes.

File: pkg/deploy/deploy_rp.go
Comment (nilsanderselde) [reply pkg/deploy/deploy_rp.go:104]: Added the check and sleep

File: pkg/util/azureclient/mgmt/compute/virtualmachinesscalesets_addons.go
Comment (nilsanderselde) [reply pkg/util/azureclient/mgmt/compute/virtualmachinesscalesets_addons.go]: Fixed

File: pkg/deploy/deploy_rp.go
Comment (m1kola) [inline pkg/deploy/deploy_rp.go:107]: Sounds like a job for `wait.PollImmediate`

Comment (nilsanderselde) [reply pkg/deploy/deploy_rp.go:107]: You're right, that is the pattern we use. I'll update it.

File: pkg/deploy/deploy_rp.go
Comment (bennerv) [inline pkg/deploy/deploy_rp.go:118]: There's a `const rpVMSSPrefix` defined in `upgrade_rp.go` within the same package which should be used instead of redefining `"rp-vmss-"`

**Review by bennerv — CHANGES_REQUESTED** [review]
We'll need to have this same functionality for the gateway VMSS once the gateway PR is merged.

File: pkg/deploy/deploy_rp.go
Comment (nilsanderselde) [reply pkg/deploy/deploy_rp.go]: Fixed

**Comment by jewzaam** [comment]
Updated labels, need this to help in https://github.com/Azure/ARO-RP/pull/1492 rollout as it'll end up with failed VMSS in every region because of how keyvault and certs rollout.

File: pkg/deploy/deploy_gateway.go
Comment (troy0820) [inline pkg/deploy/deploy_gateway.go:73]: Any reason this is 2 and the RP deploy is 3?

File: pkg/deploy/deploy_gateway.go
Comment (jewzaam) [inline pkg/deploy/deploy_gateway.go:71]:
```suggestion
```

**Review by jewzaam — CHANGES_REQUESTED** [review]

File: pkg/deploy/deploy_gateway.go
Comment (jewzaam) [inline pkg/deploy/deploy_gateway.go:71]: This is done in the loop that follows.

```suggestion
```

**Review by jewzaam — CHANGES_REQUESTED** [review]

File: pkg/deploy/deploy_gateway.go
Comment (jewzaam) [reply pkg/deploy/deploy_gateway.go:73]: It is not handling the "DeploymentFailed" code.  Maybe it should?  If so, this could be a util and reused.

**Review by bennerv — CHANGES_REQUESTED** [review]

File: pkg/deploy/deployer.go
Comment (nilsanderselde) [reply pkg/deploy/deployer.go]: fixed

File: pkg/deploy/deployer.go
Comment (nilsanderselde) [reply pkg/deploy/deployer.go]: fixed

File: pkg/deploy/deploy_gateway.go
Comment (nilsanderselde) [reply pkg/deploy/deploy_gateway.go]: fixed

File: pkg/deploy/deploy_gateway.go
Comment (nilsanderselde) [reply pkg/deploy/deploy_gateway.go:64]: yeah I was wondering if we should put the whole thing in a function.

File: pkg/deploy/deployer.go
Comment (m1kola) [inline pkg/deploy/deployer.go:179]: I know that I suggested to use `wait.PollImmediate`, but looking at the bigger picture now I'm wondering if we actually need it. I think `d.vmss.DeleteAndWait` does wait for the VMSS to be actually deleted. Or am I wrong?

File: pkg/deploy/deployer.go
Comment (nilsanderselde) [reply pkg/deploy/deployer.go:155]: I can add one

File: pkg/deploy/deployer.go
Comment (nilsanderselde) [reply pkg/deploy/deployer.go:179]: Yeah, DeleteAndWait should be enough. It's pointless to double-check that it's gone.

File: pkg/deploy/deploy_gateway.go
Comment (nilsanderselde) [reply pkg/deploy/deploy_gateway.go]: refactored into its own function

File: pkg/deploy/deployer.go
Comment (nilsanderselde) [reply pkg/deploy/deployer.go]: removed the redundant check

File: pkg/deploy/deployer.go
Comment (m1kola) [inline pkg/deploy/deployer.go:199]: Is it going to be valuable to use `Warn` here instead?

Comment (m1kola) [reply pkg/deploy/deployer.go:199]: Also a nit: maybe to decrese nesting exit early if provisioning state is not failed?

```go
if *vmss.ProvisioningState != string(mgmtcompute.ProvisioningStateFailed) {
	return false // should it be true, actually?
}
```

File: pkg/deploy/deployer.go
Comment (m1kola) [inline pkg/deploy/deployer.go:205]: We need true here, I think?

File: pkg/deploy/deployer.go
Comment (m1kola) [inline pkg/deploy/deployer.go:174]: Don't we need to call `removeFailedScaleset` if we get `DeploymentFailed`?

**Review by m1kola — COMMENTED** [review]
Few questions.

File: pkg/deploy/deployer.go
Comment (nilsanderselde) [reply pkg/deploy/deployer.go:175]: I don't think so, not if it's a spurious error as the comment states.

File: pkg/deploy/deployer.go
Comment (nilsanderselde) [reply pkg/deploy/deployer.go:205]: If the VMSS exists and was not in Failed provisioning state, we return `false` here, since false means "not deleted". At this point in the code, we would not have deleted it.

File: pkg/deploy/deployer.go
Comment (nilsanderselde) [reply pkg/deploy/deployer.go:200]: I'll switch to `Warn` and decrease the nesting.

Comment (nilsanderselde) [reply pkg/deploy/deployer.go]: Done

File: pkg/deploy/deployer.go
Comment (nilsanderselde) [reply pkg/deploy/deployer.go]: done

File: pkg/deploy/deployer.go
Comment (nilsanderselde) [reply pkg/deploy/deployer.go]: If it wasn't a spurious error, the VMSS should get removed on the next loop.

**Review by m1kola — APPROVED** [review]
Looks good to me.

One suggestion in tests, but it is not that important to block the PR.

File: pkg/deploy/deploy_test.go
Comment (nilsanderselde) [reply pkg/deploy/deploy_test.go:172]: I thought it might be informative to those reading the code to see that it would be an error with the message, "the number of retries has been exceeded".

**Review by jewzaam — APPROVED** [review]

File: pkg/deploy/deploy_test.go
Comment (nilsanderselde) [reply pkg/deploy/deploy_test.go]: took your suggestion and made it simpler

**Review by jewzaam — APPROVED** [review]

**Pull request merged**
